File: platform/frontend/lib/api/notifications.ts
import { apiClient } from "./client";

export interface AppNotification {
  id: number;
  type: string;
  title: string;
  body: string | null;
  data: Record<string, unknown>;
  read_at: string | null;
  created_at: string;
}

export interface NotificationPreferences {
  in_app: boolean;
  email: boolean;
  discord_webhook_url: string | null;
  slack_webhook_url: string | null;
  event_filter: Record<string, boolean>;
}

export async function getNotifications(unreadOnly = false, limit = 50): Promise<AppNotification[]> {
  const { data } = await apiClient.get<AppNotification[]>("/notifications/", {
    params: { unread_only: unreadOnly, limit },
  });
  return data;
}

export async function getUnreadCount(): Promise<number> {
  const { data } = await apiClient.get<{ count: number }>("/notifications/unread-count");
  return data.count;
}

export async function markRead(id: number): Promise<void> {
  await apiClient.post(`/notifications/${id}/read`);
}

export async function markAllRead(): Promise<number> {
  const { data } = await apiClient.post<{ marked: number }>("/notifications/read-all");
  return data.marked;
}

export async function getNotificationPreferences(): Promise<NotificationPreferences> {
  const { data } = await apiClient.get<NotificationPreferences>("/notifications/preferences");
  return data;
}

File: platform/backend/app/api/notifications.py
from datetime import datetime
from typing import Any, Optional

from fastapi import APIRouter, Depends, Query
from pydantic import BaseModel
from sqlalchemy.orm import Session

from app.api.deps import get_current_user, get_db
from app.models.user import User
from app.services.notification_service import (
    get_preferences, list_notifications, mark_all_read, mark_read,
    unread_count, update_preferences,
)

router = APIRouter(prefix="/notifications", tags=["notifications"])


class NotificationOut(BaseModel):
    id: int
    type: str
    title: str
    body: Optional[str]
    data: dict[str, Any]
    read_at: Optional[datetime]
    created_at: datetime

    model_config = {"from_attributes": True}


class PreferencesOut(BaseModel):
    in_app: bool
    email: bool
    discord_webhook_url: Optional[str]
    slack_webhook_url: Optional[str]
    event_filter: dict[str, Any]

    model_config = {"from_attributes": True}


class PreferencesUpdateRequest(BaseModel):
    in_app: Optional[bool] = None
    email: Optional[bool] = None
    discord_webhook_url: Optional[str] = None
    slack_webhook_url: Optional[str] = None
    event_filter: Optional[dict[str, Any]] = None


class UnreadCountOut(BaseModel):
    count: int


class ReadResultOut(BaseModel):
    ok: bool


class MarkedCountOut(BaseModel):
    marked: int


@router.get("/", response_model=list[NotificationOut])
def list_notifications_endpoint(
    unread_only: bool = Query(False),
    limit: int = Query(50, ge=1, le=200),
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> list[NotificationOut]:
    return list_notifications(db, current_user.id, unread_only=unread_only, limit=limit)


@router.get("/unread-count", response_model=UnreadCountOut)
def get_unread_count(
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> UnreadCountOut:
    return UnreadCountOut(count=unread_count(db, current_user.id))


@router.post("/{notification_id}/read", response_model=ReadResultOut)
def mark_notification_read(
    notification_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> ReadResultOut:
    result = mark_read(db, current_user.id, notification_id)
    return ReadResultOut(ok=result is not None)


@router.post("/read-all", response_model=MarkedCountOut)
def mark_all_read_endpoint(
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> MarkedCountOut:
    count = mark_all_read(db, current_user.id)
    return MarkedCountOut(marked=count)


@router.get("/preferences", response_model=PreferencesOut)
def get_prefs(
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> PreferencesOut:
    return get_preferences(db, current_user.id)


@router.patch("/preferences", response_model=PreferencesOut)
def update_prefs(
    body: PreferencesUpdateRequest,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> PreferencesOut:
    return update_preferences(
        db,
        user_id=current_user.id,
        in_app=body.in_app,
        email=body.email,
        discord_webhook_url=body.discord_webhook_url,
        slack_webhook_url=body.slack_webhook_url,
        event_filter=body.event_filter,
    )
